File: smash-ranks-android/app/src/main/java/com/garpr/android/misc/IdentityManagerImpl.java
package com.garpr.android.misc;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.text.TextUtils;

import com.garpr.android.models.AbsPlayer;
import com.garpr.android.preferences.Preference;

import java.lang.ref.WeakReference;
import java.util.Iterator;
import java.util.LinkedList;
import java.util.List;

public class IdentityManagerImpl implements IdentityManager {

    private static final String TAG = "IdentityManagerImpl";

    private final List<WeakReference<OnIdentityChangeListener>> mListeners;
    private final Preference<AbsPlayer> mIdentity;
    private final Timber mTimber;


    public IdentityManagerImpl(@NonNull final Preference<AbsPlayer> identity,
            @NonNull final Timber timber) {
        mListeners = new LinkedList<>();
        mIdentity = identity;
        mTimber = timber;
    }

    @Override
    public void addListener(@NonNull final OnIdentityChangeListener listener) {
        synchronized (mListeners) {
            boolean addListener = true;
            final Iterator<WeakReference<OnIdentityChangeListener>> iterator = mListeners.iterator();

            while (iterator.hasNext()) {
                final WeakReference<OnIdentityChangeListener> reference = iterator.next();
                final OnIdentityChangeListener item = reference.get();

                if (item == null) {
                    iterator.remove();
                } else if (item == listener) {
                    addListener = false;
                }
            }

            if (addListener) {
                mListeners.add(new WeakReference<>(listener));
            }
        }
    }

    @Nullable
    @Override
    public AbsPlayer getIdentity() {
        return mIdentity.get();
    }

    @NonNull
    private String getPlayerString(@Nullable final AbsPlayer player) {
        if (player == null) {
            return "null";
        } else {
            return "(id:" + player.getId() + ") (name:" + player.getName() + ")";
        }
    }

    @Override
    public boolean hasIdentity() {
        return mIdentity.exists();
    }

    @Override
    public boolean isId(@Nullable final String id) {
        if (TextUtils.isEmpty(id)) {
            return false;
        }

        final AbsPlayer identity = getIdentity();
        return identity != null && identity.getId().equals(id);
    }

    @Override
    public boolean isPlayer(@Nullable final AbsPlayer player) {
        if (player == null) {
            return false;
        }

        final AbsPlayer identity = getIdentity();
        return identity != null && identity.equals(player);
    }

    private void notifyListeners() {
        synchronized (mListeners) {
            final Iterator<WeakReference<OnIdentityChangeListener>> iterator = mListeners.iterator();

            while (iterator.hasNext()) {
                final WeakReference<OnIdentityChangeListener> reference = iterator.next();
                final OnIdentityChangeListener item = reference.get();

                if (item == null) {
                    iterator.remove();
                } else {
                    item.onIdentityChange(this);
                }
            }
        }
    }

    @Override
    public void removeListener(@Nullable final OnIdentityChangeListener listener) {
        synchronized (mListeners) {
            final Iterator<WeakReference<OnIdentityChangeListener>> iterator = mListeners.iterator();

            while (iterator.hasNext()) {
                final WeakReference<OnIdentityChangeListener> reference = iterator.next();
                final OnIdentityChangeListener item = reference.get();

                if (item == null || item == listener) {
                    iterator.remove();
                }
            }
        }
    }

    @Override
    public void setIdentity(@Nullable final AbsPlayer player) {
        mTimber.d(TAG, "old identity is \"" + getPlayerString(getIdentity()) + "\"" +
                ", new identity is \"" + getPlayerString(player) + "\"");

        mIdentity.set(player);
        notifyListeners();
    }

}

File: smash-ranks-android/app/src/main/java/com/garpr/android/misc/IdentityManagerImpl.kt
package com.garpr.android.misc

import com.garpr.android.misc.IdentityManager.OnIdentityChangeListener
import com.garpr.android.models.AbsPlayer
import com.garpr.android.models.FavoritePlayer
import com.garpr.android.models.Region
import com.garpr.android.preferences.Preference
import java.lang.ref.WeakReference

class IdentityManagerImpl(
        private val identityPreference: Preference<FavoritePlayer>,
        private val timber: Timber
) : IdentityManager {

    private val listeners = mutableListOf<WeakReference<OnIdentityChangeListener>>()


    companion object {
        private const val TAG = "IdentityManagerImpl"
    }

    override fun addListener(listener: OnIdentityChangeListener) {
        synchronized (listeners) {
            var addListener = true
            val iterator = listeners.iterator()

            while (iterator.hasNext()) {
                val reference = iterator.next()
                val item = reference.get()

                if (item == null) {
                    iterator.remove()
                } else if (item == listener) {
                    addListener = false
                }
            }

            if (addListener) {
                listeners.add(WeakReference(listener))
            }
        }
    }

    override val identity: FavoritePlayer?
        get() = identityPreference.get()

    private fun getPlayerString(player: AbsPlayer?): String {
        return if (player == null) { "null" } else { "(id:${player.id}) (name:${player.name})" }
    }

    override val hasIdentity: Boolean
        get() = identityPreference.exists

    override fun isPlayer(player: AbsPlayer?): Boolean {
        return player != null && identity == player
    }

    override fun isPlayer(id: String?): Boolean {
        return identity?.id?.equals(id, ignoreCase = true) == true
    }

    private fun notifyListeners() {
        synchronized (listeners) {
            val iterator = listeners.iterator()

            while (iterator.hasNext()) {
                val reference = iterator.next()
                val item = reference.get()

                if (item == null) {
                    iterator.remove()
                } else {
                    item.onIdentityChange(this)
                }
            }
        }
    }

    override fun removeIdentity() {
        timber.d(TAG, "identity is being removed, old identity was \"" +
                getPlayerString(identity) + "\"")

        identityPreference.delete()
        notifyListeners()
    }

    override fun removeListener(listener: OnIdentityChangeListener?) {
        synchronized (listeners) {
            val iterator = listeners.iterator()

            while (iterator.hasNext()) {
                val reference = iterator.next()
                val item = reference.get()

                if (item == null || item == listener) {
                    iterator.remove()
                }
            }
        }
    }

    override fun setIdentity(player: AbsPlayer, region: Region) {
        timber.d(TAG, "old identity was \"" + getPlayerString(identity) + "\"" +
                ", new identity is \"" + getPlayerString(player) + "\"")

        identityPreference.set(FavoritePlayer(player.id, player.name, region))
        notifyListeners()
    }

}
